runtime-sdk/src/context.rs

```diff
@@ -54,55 +165,114 @@ pub struct DispatchContext<'a> {
-    pub fn runtime_header(&self) -> &roothash::Header {
-        self.runtime_header
+    /// Executes a function with the transaction-specific context set.
+    pub fn with_tx<F, R>(&mut self, tx: transaction::Transaction, f: F) -> R
+    where
+        F: FnOnce(TxContext<'_, '_>, transaction::Call) -> R,
+    {
+        // Create a store wrapped by an overlay store so we can either rollback or commit.
+        let store = storage::OverlayStore::new(&mut self.runtime_storage);
+
+        let tx_ctx = TxContext {
+            mode: self.mode,
+            runtime_header: self.runtime_header,
+            runtime_round_results: self.runtime_round_results,
+            consensus_state: self.consensus_state,
+            store,
+            io_ctx: self.io_ctx.clone(),
+            logger: self
+                .logger
+                .new(o!("ctx" => "transaction", "mode" => Into::<&'static str>::into(&self.mode))),
+            tx_auth_info: tx.auth_info,
+            tags: Tags::new(),
+            // NOTE: Since a limit is enforced (which is a u32) this cast is always safe.
+            max_messages: self.max_messages.saturating_sub(self.messages.len() as u32),
+            messages: Vec::new(),
+            values: &mut self.values,
+        };
+        f(tx_ctx, tx.call)
     }
+}
 
-    /// Results of executing the last successful runtime round.
-    pub fn runtime_round_results(&self) -> &roothash::RoundResults {
-        self.runtime_round_results
+impl<'a> Context for DispatchContext<'a> {
+    type S = storage::MKVSStore<&'a mut dyn mkvs::MKVS>;
+
+    fn get_logger(&self, module: &'static str) -> slog::Logger {
+        self.logger.new(o!("sdk_module" => module))
     }
 
-    /// Runtime state store.
-    pub fn runtime_state(&mut self) -> storage::MKVSStore<&mut dyn mkvs::MKVS> {
-        storage::MKVSStore::new(self.io_ctx.clone(), &mut self.runtime_storage)
+    fn mode(&self) -> Mode {
+        self.mode
+    }
+
+    fn runtime_header(&self) -> &roothash::Header {
+        &self.runtime_header
+    }
+
+    fn runtime_round_results(&self) -> &roothash::RoundResults {
+        &self.runtime_round_results
+    }
+
+    fn runtime_state(&mut self) -> &mut Self::S {
+        &mut self.runtime_storage
+    }
+
+    fn consensus_state(&self) -> &consensus::state::ConsensusState {
+        &self.consensus_state
+    }
+
+    fn tx_auth_info(&self) -> Option<&transaction::AuthInfo> {
+        None
     }
 
-    /// Emits runtime messages
-    pub fn emit_messages(&mut self, mut msgs: Vec<roothash::Message>) -> Result<(), Error> {
+    fn emit_event<E: Event>(&mut self, event: E) {
+        self.block_tags.push(event.to_tag());
+    }
+
+    fn emit_message(
+        &mut self,
+        msg: roothash::Message,
+        hook: MessageEventHookInvocation,
+    ) -> Result<(), Error> {
         // Check against maximum number of messages that can be emitted per round.
         if self.messages.len() > self.max_messages as usize {
```

Comment:
Since we are adding a new message this should actually be:
```suggestion
        if self.messages.len() >= self.max_messages as usize {
```
Also would be cool to have a test for this.

Reply:
Opened issue https://github.com/oasisprotocol/oasis-sdk/issues/112